### PR TITLE
8267844: Replace Integer/Long.valueOf() with Integer/Long.parse*() where applicable

### PR DESCRIPTION
**Pull request by stsypanov**
The code in `Integer.decode()` and `Long.decode()` might allocate two instances of Integer/Long for the negative values less than -127:
```java
Integer result;

result = Integer.valueOf(nm.substring(index), radix);
result = negative ? Integer.valueOf(-result.intValue()) : result;
```
To avoid this we can declare 'result' as `int` and use `Integer.parseInt()` method. Same applicable for `Long` and some other classes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267844](https://bugs.openjdk.java.net/browse/JDK-8267844): Replace Integer/Long.valueOf() with Integer/Long.parse*() where applicable

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5068/head:pull/5068` \
`$ git checkout pull/5068`

Update a local copy of the PR: \
`$ git checkout pull/5068` \
`$ git pull https://git.openjdk.java.net/jdk pull/5068/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5068`

View PR using the GUI difftool: \
`$ git pr show -t 5068`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5068.diff">https://git.openjdk.java.net/jdk/pull/5068.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/Integer.java [hunk 1]
```diff
@@ -1447,15 +1447,15 @@ else if (nm.startsWith("0", index) && nm.length() > 1 + index) {
             throw new NumberFormatException("Sign character in wrong position");
 
         try {
-            result = Integer.valueOf(nm.substring(index), radix);
-            result = negative ? Integer.valueOf(-result.intValue()) : result;
+            result = parseInt(nm.substring(index), radix);
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/Integer.java:1450]: Possibly a follow-up, but I think using `parseInt/-Long(nm, index, nm.length(), radix)` could give an additional speed-up in these cases (by avoiding a substring allocation).

**Review by cl4es — APPROVED** [review]
Looks fine to me.

Could you consider adding microbenchmarks for Integer/Long.decode?

File: src/java.base/share/classes/java/lang/Integer.java (hunk 1 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/lang/Integer.java:1450]: Good point! Let me check this.

Comment (stsypanov) [reply src/java.base/share/classes/java/lang/Integer.java:1450]: Indeed, looks like getting rid of `substring()` call makes things faster:
```
before

Benchmark                    (size)  Mode  Cnt   Score   Error  Units
Integers.decode                 500  avgt   15  11.444 ? 0.949  us/op
Integers.parseInt               500  avgt   15   8.669 ? 0.152  us/op
Integers.toStringBig            500  avgt   15  10.295 ? 0.612  us/op
Integers.toStringSmall          500  avgt   15   7.020 ? 0.581  us/op

Benchmark                    (size)  Mode  Cnt   Score   Error  Units
Longs.decode                    500  avgt   15  29.568 ? 9.785  us/op
Longs.repetitiveSubtraction     500  avgt   15   0.820 ? 0.153  us/op
Longs.toStringBig               500  avgt   15  13.418 ? 0.744  us/op
Longs.toStringSmall             500  avgt   15   8.180 ? 0.780  us/op


after

Benchmark                    (size)  Mode  Cnt   Score   Error  Units
Integers.decode                 500  avgt   15   7.377 ? 0.040  us/op
Integers.parseInt               500  avgt   15   8.720 ? 0.230  us/op
Integers.toStringBig            500  avgt   15  10.328 ? 0.266  us/op
Integers.toStringSmall          500  avgt   15   6.913 ? 0.178  us/op

Benchmark                    (size)  Mode  Cnt   Score   Error  Units
Longs.decode                    500  avgt   15   8.373 ? 0.708  us/op
Longs.repetitiveSubtraction     500  avgt   15   0.771 ? 0.003  us/op
Longs.toStringBig               500  avgt   15  13.126 ? 0.079  us/op
Longs.toStringSmall             500  avgt   15   6.915 ? 0.259  us/op
```

**Review by cl4es — APPROVED** [review]
Nice!

**Comment by SkateScout** [comment]
Hi,
i check Long.java line 1301...1311 and i am not sure if this code is really good.
1) If negative  is false the whole part ends with two times the same substring and this implies the same error.
2) If negative is true and we get an error than we can throw the error without an second parse step or we can use the substring from the first round.
3) Also as mentioned above the parseLong(text,radix) should be changed to parseLong(seq, beginIndex, endIndex, radix)
  this would avoid at least in the positive case the substring at all.
4) The same points are with Integer as well. 
        try {
            result = parseLong(nm.substring(index), radix);
            result = negative ? -result : result;
        } catch (NumberFormatException e) {
            // If number is Long.MIN_VALUE, we'll end up here. The next line
            // handles this case, and causes any genuine format error to be
            // rethrown.
            String constant = negative ? ("-" + nm.substring(index))
                                       : nm.substring(index);
            result = parseLong(constant, radix);
        }

**Comment by cl4es** [comment]
> Hi,
> i check Long.java line 1301...1311 and i am not sure if this code is really good.
> 
>     1. If negative  is false the whole part ends with two times the same substring and this implies the same error.
> 
>     2. If negative is true and we get an error than we can throw the error without an second parse step or we can use the substring from the first round.
> 
>     3. Also as mentioned above the parseLong(text,radix) should be changed to parseLong(seq, beginIndex, endIndex, radix)
>        this would avoid at least in the positive case the substring at all.
> 
>     4. The same points are with Integer as well.
>        try {
>        result = parseLong(nm.substring(index), radix);
>        result = negative ? -result : result;
>        } catch (NumberFormatException e) {
>        // If number is Long.MIN_VALUE, we'll end up here. The next line
>        // handles this case, and causes any genuine format error to be
>        // rethrown.
>        String constant = negative ? ("-" + nm.substring(index))
>        : nm.substring(index);
>        result = parseLong(constant, radix);
>        }

The pre-existing logic here is iffy, but I think it is correct. 

For Integer: If negative is true, then parsing "2147483648" (Integer.MAX_VALUE + 1) would throw, be reparsed as "-2147483648" and then be accepted as Integer.MIN_VALUE. This is the only case that should be non-exceptional, but it should also be exceedingly rare in practice. For negative values it improves the error messages a bit to add the "-" and reparse. 

Improving the catch clauses with `parseLong(CharSequence, int, int, int)` and adding an `if (!negative) throw e` case to the catch could theoretically improve performance of parsing the MIN_VALUE edge case and repeated decoding of malformed positive numbers, but these are rare or exceptional cases where we should favor simplicity over optimal performance

**Comment by SkateScout** [comment]
OK even if we keep out the edge case in the try block the "parseLong(nm.substring(index), radix)" could be replaced as already mentioned with parseLong(nm. index, nm.length(), radix)
and in the catch block  the idea to throw an "nice" error can be misleading.
since -02147483648 for example would become -2147483648 because the radix is 8. 
Since per Radix only one String is possible to get through would if not be faster and more clear to check (compare) if it is the matching string and return the correct value else throw the error. This is also easy to read and even if is on the edge avoid substring , concationation and reparsing.

**Comment by cl4es** [comment]
> OK even if we keep out the edge case in the try block the "parseLong(nm.substring(index), radix)" could be replaced as already mentioned with parseLong(nm. index, nm.length(), radix)

I think it already was: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Long.java#L1287

> and in the catch block the idea to throw an "nice" error can be misleading.
> since -02147483648 for example would become -2147483648 because the radix is 8.

Yes, it's slightly misleading that the radix specifier gets cropped out, but the error message does include the radix information so it's not a bug technically:
```jshell> Integer.decode("-01234567890123");
|  Exception java.lang.NumberFormatException: For input string: "-1234567890123" under radix 8```

> Since per Radix only one String is possible to get through would if not be faster and more clear to check (compare) if it is the matching string and return the correct value else throw the error. This is also easy to read and even if is on the edge avoid substring , concationation and reparsing.

It might be a bit faster for that one non-exceptional accepted input, sure. It could also incur a cost on the fast path due increasing the weight of the code. You'd need to carefully measure that the added logic and checks doesn't cause any trouble elsewhere.

**Pull request closed without merging**